Draft DDlogServer API

remysucre opened the pull request:
Creating this PR to have the DDlogServer API code reviewed. The relevant files are `rust/template/channel.rs` and `rust/template/server.rs`; other files contain tests that manually hook up DDlog instances via TCP. The tests do not yet use the API.

ryzhyk commented on rust/template/Cargo.toml:
```diff
@@ -17,10 +17,11 @@ path = "./cmd_parser"
 path = "./ovsdb"
 
 [dependencies]
+fallible-iterator = "0.2"
```
Are we still using `FallibleIterator`?

remysucre replied:
I'll change to `FallibleIterator` once the API comes into shape, trying to keep it simple for now.

ryzhyk replied:
I'm not sure it's needed given the current design philosophy to only push data that is ready in the producer.

remysucre replied:
I see!

ryzhyk commented on rust/template/channel.rs:
```diff
@@ -0,0 +1,32 @@
+// a producer
+pub trait Observer<T, E>
+{
+    fn on_start(&self) -> Response<()>;
```
How is the producer supposed to handle errors from this interface? E.g., should they stop sending data?

remysucre replied:
There's another method `on_error` from `Observer` that I'll write once the other parts come into shape.

remysucre replied:
Added `on_error` to `Observer`.

remysucre commented:
I've refactored `subscribe` to take `Arc` instead of `Box`, and added a test driver in [/test/datalog_tests/channel](https://github.com/remysucre/differential-datalog/blob/outlet/test/datalog_tests/channel/src/main.rs). If the server API looks fine, I'll try to implement a TCP channel next.

remysucre commented on test/datalog_tests/channel/Cargo.toml:
```diff
@@ -0,0 +1,11 @@
+[package]
```
I'll move this to `rust/template/channel` once it's working. To test it, I can have a main that imports the channel library.

remysucre replied:
Moved to `rust/template/tcp_channel`.

remysucre commented on rust/template/server.rs:
```diff
@@ -0,0 +1,129 @@
+
+pub struct Outlet
+{
+    tables: HashSet<RelId>,
```
It's a subset of the server's tables, so there shouldn't be duplicates.

remysucre commented on rust/template/channel.rs:
```diff
@@ -0,0 +1,32 @@
+
+// The consumer can subscribe to the channel
+// which acts as an observable of deltas.
+pub trait Observable<T, E>
```
Yes

remysucre replied:
Now `Observer` methods return `Result<(), E>`.

remysucre commented on rust/template/channel.rs:
```diff
@@ -0,0 +1,32 @@
+pub trait Observable<T, E>
+where T: Send, E:Send
+{
+    fn subscribe<'a>(&'a mut self, observer: Arc<Mutex<dyn Observer<T, E>>>) -> Box<dyn Subscription + 'a>;
```
Good point, I can try that

remysucre replied:
Now `subscribe` takes just an `Arc`, and `Observer` methods only need `&self`. I will explore pushing the `Arc` to the `Observer` side later.

remysucre replied:
Now `Arc` is inside the `Observer` side.

remysucre commented on rust/template/channel.rs:
```diff
@@ -0,0 +1,32 @@
+
+// Stop listening to changes from the observable
+pub trait Subscription<'a> {
+    fn unsubscribe(&'a mut self);
```
Yes you're right

remysucre commented on rust/template/server.rs:
```diff
@@ -0,0 +1,129 @@
+        for outlet in &mut self.outlets {
+            if let Some(ref observer) = outlet.observer {
+                let upds = outlet.tables.iter().flat_map(|table| {
+                    changes.as_ref().get(table).unwrap().iter().map(move |(val, weight)| {
```
I can change `add_stream` to validate the id's.

remysucre replied:
I guess it's still unsafe even if we validate the id's to exist in the table in `add_stream`, since some of the streamed tables might not have changed.

remysucre replied:
I changed `add_stream` to validate the id's. When the id exists in the db but not in the delta, I ignore them when passing along the delta.

remysucre replied:
Hmm there isn't actually a way to validate the ID's unless `do_store` is set to `true`. So for now I just ignore invalid ID's.

remysucre commented on rust/template/server.rs:
```diff
@@ -0,0 +1,129 @@
+use std::fmt;
+
+pub struct UpdatesSubscription<'a> {
+    observer: &'a mut Option<Arc<Mutex<dyn Observer<Update<super::Value>, String>>>>,
```
Actually, this `Option` comes from `Outlet`. An `Outlet` has no observer until one subscribes to it.

remysucre commented on rust/template/channel.rs:
```diff
@@ -0,0 +1,32 @@
+use differential_datalog::program::Response;
```
I moved the trait declaration to its own crate. What about the implementation in `server.rs`? I feel that's more closely tied to the rest of DDlog template

remysucre commented on rust/template/server.rs:
```diff
@@ -0,0 +1,129 @@
+pub struct DDlogServer
+{
+    prog: HDDlog,
+    outlets: Vec<Outlet>,
```
Turns out there's a deeper problem. `add_stream` mutably borrows `self`, which means there can never be more than one outlet at a time.

remysucre replied:
I wrapped the `Outlet`s in `Arc<Mutex<>>` to allow multiple of them coexist, and added `remove_stream`.

remysucre commented on rust/template/server.rs:
```diff
@@ -0,0 +1,129 @@
+        self.prog.apply_valupdates(updates.map(|upd| match upd {
+            Update::Insert{relid: relid, v: v} =>
+                Update::Insert{
+                    relid: *self.redirect.get(&relid).unwrap(),
```
Now when the relid is not in the redirect, I just use the original ID. I think this can be a reasonable semantics (the user only specify the ID's they want to redirect, and leave the rest alone).
